### PR TITLE
[TensorExpr] Extend arithmetic simplifier to work with multi variable expressions (Attempt 2)

### PR DESCRIPTION
**Pull request by nickgg**
https://github.com/pytorch/pytorch/pull/35127 was landed and reverted because I missed a test fail (oops). I have found and fixed the issue, which was due to zero terms being introduced after the point that filtered them out (usually required NAN/INF, e.g. x / INF => 0).

See https://github.com/pytorch/pytorch/pull/35127 for more info.

### CODE REVIEW

**Review by ZolotukhinM — APPROVED** [review]
Approving since it's a fix to an already approved PR.

**Comment by nickgg** [comment]
OK have rebased back on the various commits, this is the one!

File: test/cpp/tensorexpr/tests.h [hunk 1]
```diff
@@ -244,7 +257,7 @@ TH_FORALL_TENSOREXPR_TESTS_LLVM(DECLARE_TENSOREXPR_TEST)
 #ifdef USE_CUDA
 TH_FORALL_TENSOREXPR_TESTS_CUDA(DECLARE_TENSOREXPR_TEST)
 #endif
-#undef DECLARE_TENSOREXPR_TEST
+#undef DECLARE_TEST
```
Comment (ZolotukhinM) [inline test/cpp/tensorexpr/tests.h]: An undesired side-effect of conflict resolution, please undo this change.

File: tools/build_variables.bzl [hunk 2]
```diff
@@ -207,6 +207,7 @@ libtorch_sources = [
     "torch/csrc/jit/tensorexpr/ir.cpp",
     "torch/csrc/jit/tensorexpr/ir_mutator.cpp",
     "torch/csrc/jit/tensorexpr/ir_printer.cpp",
+    "torch/csrc/jit/tensorexpr/ir_simplifier.cpp",
```
Comment (ZolotukhinM) [inline tools/build_variables.bzl]: Please add `hash_provider.cpp` as well.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 3]
```diff
@@ -0,0 +1,1093 @@
...
+  return hash;
+}
+
+void Polynomial::sort() {
+  if (dtype().is_floating_point()) {
+    throw std::logic_error("reordering FP ops");
+  }
+  std::sort(
+      variables_.begin(), variables_.end(), [&](const Expr* a, const Expr* b) {
+        return hasher_.hash(a) < hasher_.hash(b);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: It is generally a bad idea to sort on hash, as we've already discovered - it can make tests brittle and can in future lead to non-deterministic behavior (even if it's not like that now). I suggest following up on that - a better way would be to normalize/canonicalize the expression - i.e. define a strict order how sub-expressions should be placed in the parent expression.

**Review by ZolotukhinM — APPROVED** [review]
Approving as already reviewed. Please fix a couple of places (see inline) before landing. Also, I suggest splitting big PRs into a stack of smaller PRs in future. Thanks!

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp (hunk 3 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: The hash defines an ordering though, doesn't it. The issues we've been having in this diff is that the hash wasn't consistent across platforms.

I'm a bit resistant to some kind of adhoc ordering because in the case of compound opaque objects I think it can get complicated. Am open to suggestions on this though.

Comment (nickgg) [reply torch/csrc/jit/tensorexpr/ir_simplifier.cpp]: You're not wrong though, I found a bug in the hash and it changes the order of a bunch of things.
